### ISSUE TITLE
dependency notation does not support maven version range syntax containing `)` and `,`

### ISSUE THREAD

**Issue by NikkyAI** [holgerbrandl/kscript#166]
EDIT: Maven supports version ranges natively: examples: http://www.mojohaus.org/versions-maven-plugin/examples/resolve-ranges.html
an it works in kscript once the annotation and comment parsing is not messing up on `,` and `)`

old issue content:

i am writing a library and tool that i try to make compatilble / consumable by kscript
the versions are pushed to maven by jenkins , each git commit increments the build number

ideally i would like to use
```
@file:DependsOnMaven("moe.nikky.voodoo:dsl:0.4.0+")
```
instead of
```
@file:DependsOnMaven("moe.nikky.voodoo:dsl:0.4.0-132")
```

**Comment by NikkyAI** [comment]
related: https://youtrack.jetbrains.com/issue/KT-27051

**Comment by holgerbrandl** [comment]
Could you point me to some reference docs about dynamic versions? What is the format? Is it "just"  about supporting dashes and pluses in the version field of the artifact locator?

**Comment by NikkyAI** [comment]
gradle docs are not very helpful: https://docs.gradle.org/current/userguide/declaring_dependencies.html#sub:declaring_dependency_with_dynamic_version
seems like they have a open issue about documenting the syntax: https://github.com/gradle/gradle/issues/5603

**Comment by holgerbrandl** [comment]
Why don't you use jitpack to get temporary artifacts?

**Comment by NikkyAI** [comment]
i did that but it kept reusing the same -SNAPSHOT version that maven cached and update that only somewhat randomly
aalso jitpack had issues buildingthe software, took way too long compared to a friends jenkins and maven repo

**Comment by holgerbrandl** [comment]
Indeed for continuous rebuilding it is sometimes a bit sluggish.

Coming back to the dynamic dependencies. What is the format? Is it "just" about supporting dashes and pluses in the version field of the artifact locator?

In order to support it we would need at least some (unit-)test as the dependency lookup is likely to be replace with a different backend (kotlin-scripting-api, aether from existing PR).

**Comment by NikkyAI** [comment]
could this be used ? http://ant.apache.org/ivy/history/latest-milestone/ivyfile/dependency.html

**Comment by holgerbrandl** [comment]
Would maven support such versions? Since it is the current resolver backend, it would be hard to implement it if not. Or we would need to change until we change to something like kotlin-main-api or aether (see PR).

**Comment by NikkyAI** [comment]
well i have no experience using POMs directly, but i am currently trying to add `mvn versions:resolve-ranges`  to the maven invocation in kscript

**Comment by NikkyAI** [comment]
apparently ranges like `[1.0,2.0]` work, but open ranges eg. `[1.0,)` are breaking

**Comment by NikkyAI** [comment]
for some reason `[0.4.0,)` gets regexed into `[0.4.0,` which produces errors in maven

**Comment by NikkyAI** [comment]
so spliting versions with `,` breaks the input of maven version ranges, once i fixed that (an the tests) everything else works, i think i should PR this and continue the discussion there

**Comment by holgerbrandl** [comment]
Thanks for the PR. I guess we can not simply change the delimiter from , to ; because of backwar compatibility reasons. But do we have to? Can these dynamic versions actually end with a comma? Your examples seem to indicate the opposite?

**Comment by NikkyAI** [comment]
which are the lagacy usecases i should test for ?
currently the whole testsuite seems to pass

**Comment by holgerbrandl** [comment]
> is not really easy to parse, i could also just ignore it and just fix it for @DependsOn annotations
but that feels bad as well.. and you would have to warn people about what won't work in which case..

+1 It's hard to parse. +1 consistency at this level would be nice, although I'd prefer annotations as well.

So if not even idea is supporting them in their gradle/maven sync, does kscript really need/want it?

**Comment by NikkyAI** [comment]
i think idea would support it.. i mean kscript does nto let me create the project because it runs through codepaths that are broken because `//DEPS` i think
i just got a bit fed up because of that
see my last comment on the closed PR

i have a new attempt of just supporting `1.0+` and converting that into a maven open range during POM building
not having to touch the rest of the parsing (although i will include the improved annotation parsing)
and it will work in comments and annotations

**Comment by holgerbrandl** [comment]
@NikkyAI  Is is possible that dynamic versions can just be resolved when running kscript (including the patch from above), if at least one version is sitting in the local `.m2` cache already?

E.g when using some non-yet-`~/.m2` dependency such as `com.indeed:util-mmap:jar:1.0.+`, it fails to resolve them for me.

**Comment by holgerbrandl** [comment]
... which might be an argument in favor of PR #49 instead of maven?

**Comment by holgerbrandl** [comment]
Since the PR has been long merged, I'm also closing the corresponding ticket.
